### ISSUE TITLE
Request body deserialization failure when calling actor from Dapr-enabled service

### ISSUE THREAD

**Issue by philliphoff** [dapr/dotnet-sdk#138]
## Expected Behavior

Assume an actor (A1) is registered with Dapr-enabled service (S1).
Assume another Dapr-enabled service (S2) calls actor (A1) when invoked via client (C1).

The expectation is that the call to actor A1 succeeds.

## Actual Behavior

The call to actor A1 fails in service (S1) with a request body deserialization error:

```
== APP ==       Request starting HTTP/1.1 PUT http://127.0.0.1:3000/actors/CardActor/1/method/SetDetailsAsync application/json 194
== APP == ERROR: ActorManager:  Got exception from actor method invocation: Newtonsoft.Json.JsonReaderException: Unexpected character encountered while parsing value: @. Path '', line 0, position 0.
== APP ==    at Newtonsoft.Json.JsonTextReader.ParseValue()
== APP ==    at Newtonsoft.Json.JsonReader.ReadAndMoveToContent()
== APP ==    at Newtonsoft.Json.JsonReader.ReadForType(JsonContract contract, Boolean hasConverter)
== APP ==    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
== APP ==    at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
== APP ==    at Newtonsoft.Json.JsonSerializer.Deserialize(JsonReader reader, Type objectType)
== APP ==    at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
== APP ==    at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type)
== APP ==    at Dapr.Actors.Runtime.ActorManager.<>c__DisplayClass15_0.<<DispatchWithoutRemotingAsync>g__RequestFunc|0>d.MoveNext()
== APP == --- End of stack trace from previous location where exception was thrown ---
== APP ==    at Dapr.Actors.Runtime.ActorManager.DispatchInternalAsync[T](ActorId actorId, ActorMethodContext actorMethodContext, Func`3 actorFunc, CancellationToken cancellationToken)
== APP == info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
== APP ==       Request finished in 1.8537000000000001ms 500 application/json
```

> The request has the `Content-Type` set to `application/json` but the body appears to be octet-stream (?) encoded, which is what generates the JSON deserialization exception.

## Steps to Reproduce the Problem

1. Extract the attached repro solution
1. Run the `Actor` project:
  `dapr run --app-id myapp --app-port 3000 --port 3500 dotnet run`
1. Run the `Client` project (and note the HTTP port chosen by Dapr):
  `dapr run --app-id test-client --app-port 5000 dotnet run`
1. Make a call via the Dapr HTTP proxy port:
  `http://localhost:<port>/v1.0/invoke/test-client/method/test`
1. Make a call via the raw service port:
  `http://localhost:5000/test`

> The call should fail regardless of being made via the Dapr HTTP proxy port or the raw service port.  If you run the `Client` project without Dapr (e.g. `dotnet run`), then the call via the raw service port will succeed.

[dapr-test.zip](https://github.com/dapr/dotnet-sdk/files/3769631/dapr-test.zip)

**Comment by amanbha** [comment]
@philliphoff  Thanks for reporting it,  Could you please share what was your response body? 
Our samples make similar calls https://github.com/dapr/dotnet-sdk/blob/master/samples/Actor/Readme.md

**Comment by philliphoff** [comment]
@amanbha The request body (of the call being made into the actor service) looks similar to:

```
@\u000eWrappedMsgBody\b\nurn:actors\t\u0001i)http://www.w3.org/2001/XMLSchema-instance@\u0005value.\u0004type�\u0016SetDetailsAsyncReqBody@\adetails\t\u0001a7http://schemas.datacontract.org/2004/07/DaprCards.Cards^\u0006userId�\u000242^\u0005value�\u0019\u0001\u0001\u0001
```

Note that the sample makes calls into the actor service from a non-Dapr-enabled console application, which does not exhibit this behavior; to reproduce the issue, the service making the call to the actor service must, itself, be within the context of Dapr.

**Comment by amanbha** [comment]
I see, For invocation without Remoting directly over http, it  has to be json. The server side deserializes the json and uses that as the actor method argument.
https://github.com/dapr/dotnet-sdk/blob/master/docs/get-started-dapr-actor.md#invoke-actor-method-without-actor-service-remoting

**Comment by philliphoff** [comment]
@amanbha In my scenario (and the repro solution), the strongly-typed remoting API is used.  In any. case, since the solution is using the .NET SDKs on both sides, how the call is made is completely opaque to the business logic.

**Comment by amanbha** [comment]
Oh ok,. The repro steps shows invoking directly over http without remoting. I will take a look at your sample code also.

**Comment by philliphoff** [comment]
@amanbha Sorry, I probably could have been more clear.  The HTTP call is being made to a Dapr service which is then invoking an actor in *another* Dapr service (which fails).

Fails: `Client C1 --> Dapr --> Service S2 --> Dapr --> Service S1 with actor A1`

> Note: it also fails if Client C1 bypasses Dapr and calls directly to Service S2

Works: `Client C1 --> Service S2 --> Dapr --> Service S1 with actor A1`
Works: `Client C1 --> Dapr --> Service S1 with actor A1`

**Comment by amanbha** [comment]
@philliphoff  Its because because the [header](https://github.com/dapr/dotnet-sdk/blob/master/src/Dapr.Actors/Constants.cs#L14) which client side adds to signify its a remoting call is missing when it comes to the server side, so it gets treated as a [non-remoting call](https://github.com/dapr/dotnet-sdk/blob/master/src/Dapr.Actors.AspNetCore/RouterBuilderExtensions.cs#L58), let me debug it further to see who is dropping that header.

**Comment by amanbha** [comment]
Investigated further and created Daprd issue: https://github.com/dapr/dapr/issues/717

**Comment by amanbha** [comment]
Issue in Dapr runtime is fixed and will be available in 0.2.0 release which is planned for later this week.
